lib/src/shared/main/java/com/couchbase/lite/internal/core/C4Document.java

```diff
@@ -17,108 +17,88 @@
 //
 package com.couchbase.lite.internal.core;
 
-import android.support.annotation.GuardedBy;
-import android.support.annotation.NonNull;
 import android.support.annotation.VisibleForTesting;
 
 import com.couchbase.lite.LiteCoreException;
-import com.couchbase.lite.LogDomain;
-import com.couchbase.lite.internal.CouchbaseLiteInternal;
 import com.couchbase.lite.internal.fleece.FLDict;
 import com.couchbase.lite.internal.fleece.FLSharedKeys;
 import com.couchbase.lite.internal.fleece.FLSliceResult;
-import com.couchbase.lite.internal.support.Log;
-import com.couchbase.lite.internal.utils.Preconditions;
-import com.couchbase.lite.utils.Fn;
 
 
 @SuppressWarnings({"PMD.GodClass", "PMD.TooManyMethods"})
-public class C4Document {
+public class C4Document extends C4NativePeer {
     public static boolean dictContainsBlobs(FLSliceResult dict, FLSharedKeys sk) {
         return dictContainsBlobs(dict.getHandle(), sk.getHandle());
     }
 
-    //-------------------------------------------------------------------------
-    // Member Variables
-    //-------------------------------------------------------------------------
-
-    @NonNull
-    private final Object lock = new Object(); // lock for thread-safety
-
-    @GuardedBy("lock")
-    private long handle; // hold pointer to C4Document
-
-    private Exception freedAt;
-
     //-------------------------------------------------------------------------
     // Constructor
     //-------------------------------------------------------------------------
     C4Document(long db, String docID, boolean mustExist) throws LiteCoreException { this(get(db, docID, mustExist)); }
 
     C4Document(long db, long sequence) throws LiteCoreException { this(getBySequence(db, sequence)); }
 
-    C4Document(long handle) {
-        Preconditions.assertNotZero(handle, "handle");
-        this.handle = handle;
-    }
+    C4Document(long handle) { super(handle); }
 
     //-------------------------------------------------------------------------
     // public methods
     //-------------------------------------------------------------------------
 
     // - C4Document
 
-    public int getFlags() { return withHandle(C4Document::getFlags, 0); }
+    public int getFlags() { return withPeer(0, C4Document::getFlags); }
```

Comment:
This is a discussion for withPeer() method in general (not only with getFlags() method. I'm going back and forth between returning a default value or just throw an exception so that we will ensure that there is no unexpected behavior as a result of the default value returned. What do you think about throwing the exception instead?

Reply:
Yeah... That's a really good point.
I currently allow both, because I'm trying to do this incrementally.
I have defined `getPeer` which throws the exception, and `withPeer` that doesn't.
I did that because there are places in the current code that will, absolutely, fail, if the peer has been released.  There are some places, though, that seem to want default values.  This change is, already, huge.  I didn't want to add more too it.  For now, it should work exactly as the old code did.